robot/control/Src/motion-control/VelocityKalmanTuner/KalmanFilter.py

```diff
@@ -0,0 +1,195 @@
+        for i, angle in enumerate(self.wheel_angles):
+            self.H[i - 1, :] = np.array([-np.sin(angle), np.cos(angle), self.wheel_dist])
+        self.H[4, :] = np.array([0, 0, 1])
+        self.H /= -self.wheel_radius
+
+        self.Q = np.round(3.0 * self.process_noise * 0.0001 * np.eye(self.num_states), 10)
+        self.R = np.eye(self.num_outputs)
+        self.R[:4, :4] *= self.encoder_noise
+        self.R[4, 4] *= self.gyro_noise
+
+        if self.steady_state:
+            self.P = scipy.linalg.solve_discrete_are(self.A.T, self.H.T, self.Q, self.R)
+            self.K = self.P @ self.H.T @ np.linalg.inv(self.H @ self.P @ self.H.T + self.R)
+        else:
+            self.P = np.eye(self.num_states) * init_covariance
+            self.K = np.zeros((self.num_states, self.num_outputs))
+
+    def step(self, dt, verbose=False, changeable_only=False):
+        self.t += dt
+        self.prediction_step()
+        self.update_step()
+        if verbose:
+            print("-" * 40)
+            v = self.vars(changeable_only)
+            for key in v:
+                print(key, ":\n", np.round(v[key], 5))
+        return self.t, self.x_hat, self.x
+
+    def generate_u(self):
+        return np.zeros((self.num_inputs, 1))
+
+    def prediction_step(self):
+        # Control input, returns 0 for now
+        self.u = self.generate_u()
+
+        # A priori state estimate
+        # x_hat' = A * x_hat_prev + B * u
+        self.x_hat = self.A @ self.x_hat + self.B @ self.u
+
+        if not self.steady_state:
+            # A priori covariance estimate
+            # P' = A * P_prev * A.T + Q
+            self.P = self.A @ self.P @ self.A.T + self.Q
+
+    def update_step(self):
+        if not self.steady_state:
+            # Optimal Kalman Gain
+            # K = P' * H.T * (H * P' * H.T + R)^-1
+            self.K = self.P @ self.H.T @ np.linalg.inv(self.H @ self.P @ self.H.T + self.R)
+
+        # self.w = self.rng.multivariate_normal((0, 0, 0), self.Q, method='eigh').reshape((self.num_states,1))
+        if self.step_response:
+            a = 1
+            self.x = np.heaviside((self.t-a)*np.ones((self.num_states,1)), a)
+        else:
+            self.w = self.Q @ np.random.randn(self.num_states, 1)
+            self.x = self.A @ self.x + self.B @ self.u + self.w
+
+        # Observation
+        # self.v = self.rng_5.multivariate_normal((0, 0, 0, 0, 0), self.R, method='eigh').reshape((self.num_outputs,1))
+        self.v = self.R @ np.random.randn(self.num_outputs, 1)
+        self.z = self.H @ self.x + self.v
+
+        # Prefit
+        # y = z - H * x_hat
+        y = self.z - self.H @ self.x_hat
+
+        # A posteriori state estimate
+        # x_hat = x_hat' + K * (y - H * x_hat_est)
+        self.x_hat += self.K @ y
+
+        if not self.steady_state:
+            # A posteriori covariance estimate
+            # P = (I - K * H) * P'
+            self.P = (np.eye(self.num_states) - self.K @ self.H) @ self.P
```

Comment:
Missing post fit? y_k|k

Reply:
I guess it really depends if you're exposing y. Up to you

Reply:
I'm not exposing y for now, I'll add it later